tests/constraints/tests.py

```diff
@@ -68,21 +84,99 @@ def test_repr(self):
+
+    def test_model_validation_with_condition(self):
+        Product(name=self.p1.name, color='blue').validate_unique()
+        Product(name=self.p2.name).validate_unique()  # validation ignored, will raise on DB level
 
     def test_name(self):
         constraints = get_constraints(Product._meta.db_table)
-        expected_name = 'unique_name'
+        expected_name = 'name_color_uniq'
         self.assertIn(expected_name, constraints)
+
+    def test_constructor_condition_must_be_q(self):
+        with self.assertRaisesMessage(ValueError, 'UniqueConstraint.condition must be a Q instance.'):
+            models.UniqueConstraint(name='uniq', fields=['name'], condition='invalid')
+
+
+@skipIfDBFeature('supports_partial_indexes')
+class ConditionalUniqueConstraintIgnoredTests(TransactionTestCase):
+    available_apps = ['constraints']
+
+    def test_create_conditional_constraint_ignored(self):
+        constraint = models.UniqueConstraint(
+            name='name_without_color_uniq2',
+            fields=['name'],
+            condition=models.Q(color__isnull=True),
+        )
+        with connection.schema_editor() as editor:
+            self.assertIsNone(constraint.create_sql(Product, editor))
+            editor.add_constraint(Product, constraint)
+
+    def test_delete_conditional_constraint_ignored(self):
+        constraint = models.UniqueConstraint(
+            name='name_without_color_uniq',
+            fields=['name'],
+            condition=models.Q(color__isnull=True),
+        )
+        with connection.schema_editor() as editor:
+            self.assertIsNone(constraint.remove_sql(Product, editor))
+            editor.remove_constraint(Product, constraint)
+
+
+@skipUnlessDBFeature('supports_partial_indexes')
+class ConditionalUniqueConstraintTests(TransactionTestCase):
+    available_apps = ['constraints']
+
+    def test_create_conditional_constraint(self):
+        constraint = models.UniqueConstraint(
+            name='name_without_color_uniq2',
+            fields=['name'],
+            condition=models.Q(color__isnull=True),
+        )
+        with connection.schema_editor() as editor:
+            self.assertIn(
+                'WHERE %s IS NULL' % editor.quote_name('color'),
+                str(constraint.create_sql(Product, schema_editor=editor))
+            )
+            editor.add_constraint(Product, constraint)
+        self.assertIn(constraint.name, get_constraints(Product._meta.db_table))
+
+    def test_delete_conditional_constraint(self):
+        constraint = models.UniqueConstraint(
+            name='name_without_color_uniq',
+            fields=['name'],
+            condition=models.Q(color__isnull=True),
+        )
+        with connection.schema_editor() as editor:
+            self.assertIn(constraint.name, get_constraints(Product._meta.db_table))
+            editor.remove_constraint(Product, constraint)
+        self.assertNotIn(constraint.name, get_constraints(Product._meta.db_table))
```

Comment:
Unless the constraint is added back this alters the global test state. Most of the schema altering tests above are also affected by that. These tests should live in the `schema` and `migrations` test apps which has the necessary setup in place to prevent this from happening.

Have a look at #7615 and #10639 for examples.

Reply:
refactored and moved to tests/migrations/test_operations.py